### PR TITLE
C++: use TaintTracking::Configuration in DefaultTaintTracking

### PR DESCRIPTION
**Pull request by rdmarsh2**
Switch from using additional flow steps with a `DataFlow::Configuration` to using a `TaintTracking::Configuration`. This makes future improvements to `TaintTracking::Configuration` reflected in `DefaultTaintTracking` without further effort. It also removes the predictability constraint in `DefaultTaintTracking`, which increases the number of results, with both new true positives and new false positives. Those may need to be addressed on a per-query basis.

### CODE REVIEW

**Comment by jbj** [comment]
How awful would it be to provide an abstract class in `TaintTrackingUtil.qll` that `DefaultTaintTracking` can extend in order to configure the taint-tracking library to enforce the predictability constraint? That could be either a general mechanism for suppressing certain additional edges, or it could be very specific to the predictability constraint.

**Comment by rdmarsh2** [comment]
I don't like the idea of using a configuration-independent abstract class to suppress edges, since an accidental indirect import would lead to invisible false negatives. It's easier to do with a barrier in the configurations in `DefaultTaintTracking` if we just want the old behavior, and anything where the new behavior is better can eventually use a custom configuration.

**Comment by rdmarsh2** [comment]
The changes in CWE-134/semmle/argv/argvLocal.expected are the same ones that are fixed by https://github.com/github/codeql/pull/3867/commits/ad4a753daac846c8c37d5af38e0f65c2115ebbc9 in the uncontrolled format string PR

**Comment by geoffw0** [comment]
I've started SAMATE runs on this PR [here](https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp/551/) and [here](https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp2/33/) (on a copy of the branch with latest main merged in, because I wanted to be sure of including https://github.com/github/semmle-code/pull/38197).  The results could help to build confidence in this PR, but mainly I'd like to know what to expect in the next run after this is merged.

**Comment by jbj** [comment]
For the record, the initial CPP-Differences job is here: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1550/

**Comment by geoffw0** [comment]
I'm pretty happy with this, @jbj how do you feel about it?

**Comment by jbj** [comment]
> I'm pretty happy with this, @jbj how do you feel about it?

I'm waiting to hear what @rdmarsh2 says after reading through [the CPP-Differences results](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1550//artifact/changes.html).

**Comment by geoffw0** [comment]
@rdmarsh2 we're waiting for your thoughts.

Also merge conflicts have appeared, I believe caused by https://github.com/github/codeql/pull/4741.

**Comment by rdmarsh2** [comment]
There's some TaintedAllocationSize false positives involving the pointer difference operator, which I think we excluded from `DefaultTaintTracking` but not `TaintTracking::Configuration`. There's also a number of lost results that I'm still triaging, but many of them were true positives. The new non-TaintedAllocationSize results all look like true positives to me.

**Comment by rdmarsh2** [comment]
It looks like the lost results mostly have to do with double indirections that were previously handled by pointer/object conflations. I'll try and get somewhere on that issue next week.

**Comment by geoffw0** [comment]
What's the status of this PR?  From my perspective it was looking like a good change.

**Comment by MathiasVP** [comment]
I've merged in the latest `main` and resolved the conflicts. It took a couple of tries to get all of the conflicts resolved correctly (hence the force-pushes). I've started a fresh CPP-differences for this PR: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1774/

**Comment by MathiasVP** [comment]
> I've merged in the latest `main` and resolved the conflicts. It took a couple of tries to get all of the conflicts resolved correctly (hence the force-pushes). I've started a fresh CPP-differences for this PR: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1774/
> 
> Edit: I found a couple of simple ways to improve our test results in [eb4f1e1](https://github.com/github/codeql/commit/eb4f1e1ba03cdbda41df975295dbf8e5e4dcd0f7) so I've started [yet another CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1775/).

CPP-differences came back with quite a lot of new results! It seems like most of them are good results, but there are some cases of field to object taint that causes field conflation, which lead to false positives:

One example is on `OpenJDK` when running `cpp/path-injection`:

```cpp
void SystemProcessInterface::SystemProcesses::ProcessIterator::get_exe_name() {
  FILE* fp;
  char  buffer[PATH_MAX];

  jio_snprintf(buffer, PATH_MAX, "/proc/%s/stat", _entry->d_name);
  buffer[PATH_MAX - 1] = '\0';
  fgets(buffer, PATH_MAX, fp);
  char* start, *end;
  // exe-name is between the first pair of ( and )
  start = strchr(buffer, '(');
  end = strrchr(start, ')');
  size_t len;
  len = MIN2<size_t>(end - start, sizeof(_exeName) - 1);
  memcpy(_exeName, start, len);
}
```

where the `memcpy` ends up tainting `*this` instead of being a storeStep that writes to the `_exeName` field.

**Comment by MathiasVP** [comment]
bb53780 ports a couple of rules from DefaultTaintTracking:
- Flow through unary instructions (that are now `FieldAddressInstructions`)
I think this is actually a good taint rule that we should have in `TaintTrackingUtil`, so this is probably controversial.
- Conflation from `*p` to `p` for parameters.
This is actually one of the rules that https://github.com/github/codeql/pull/5089 removes, but without adding an equivalent store/read step we lose quite a few results when we remove it. So I've just ported these problematic rules from `DefaultTaintTracking` into `TaintTrackingUtil`, and we can update https://github.com/github/codeql/pull/5089 to remove these rules once we merge this PR.

Locally it looks like we regain the lost results. I've started a new [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1801/) to confirm this. If this comes back with good results I think this PR is good to go.

**Comment by MathiasVP** [comment]
> Locally it looks like we regain the lost results. I've started a new [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1801/) [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1812/) to confirm this. If this comes back with good results I think this PR is good to go.

CPP-differences looks good. All the results look like good results.

There appears to be a lot of noise: Linux has a 10% slowdown, but that appears to be due to slowdowns in `SignedOverflowCheck.ql` and `NewArrayDeleteMismatch.ql`, and neither of those use taint tracking. Similarly for `ChakraCore` and `kamailio`.

**Comment by geoffw0** [comment]
> CPP-differences looks good. All the results look like good results.

Just had a look at some of the new results, I agree they look reasonable (the only issues I found were due to existing flaws in the query and not bad taint).  I was pleased to see flow through `std::string`.

I don't think I've personally reviewed this thoroughly enough to hit 'merge' though.

**Comment by MathiasVP** [comment]
> I don't think I've personally reviewed this thoroughly enough to hit 'merge' though.

I also need to open up an internal PR. I've been waiting with that until we got some good results here, but I think it's about time I do that.

**Comment by MathiasVP** [comment]
Since you opened the PR, I guess I can't request a review from you @rdmarsh2. So please take this mention as a request for review :) The internal PR should be up-to-date.

File: cpp/ql/test/library-tests/dataflow/taint-tests/stringstream.cpp
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/taint-tests/stringstream.cpp:41]: Typo. I wonder how this comment gets interpreted.

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:195]: There's now an `ir` result, but is it from the correct source (line 192) or the wrong one (line 193)? Same in `test_mempcpy` and `test_memccpy` further down.

File: cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_sinks_only/defaulttainttracking.cpp
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_sinks_only/defaulttainttracking.cpp:22]: Is this result recoverable?

**Review by jbj — COMMENTED** [review]
LGTM except some minor comments.

I'm still lacking an overview of which tests are affected by the new change (because they use the `Configuration` you changed) and which tests aren't (because they use one of the `Configuration`s that still use data flow rather than taint tracking).

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:195]: Good point. Currently, we only allow a `line:column` annotation on a sink when there's more than one source for it, but for cases such as this, we actually want to highlight the exact source even though there should only be one.

File: cpp/ql/test/library-tests/dataflow/taint-tests/stringstream.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/stringstream.cpp:41]: Hm, 🤔 Good question! I would think it would be interpreted as `// $ ast,MISSIN,ir` could I could be wrong. Anyway, I'll fix it up.

File: cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/taint.cpp:195]: You're gonna have to trust me when I say that it's the correct one for now 😄 I've made a [PR](https://github.com/github/codeql/pull/5417) with the changes to the inline expectation library so that we can provide "optional" values. Once that PR gets merged (assuming the other languages are okay with it 🤞) we can revisit this annotation.

**Comment by MathiasVP** [comment]
> I'm still lacking an overview of which tests are affected by the new change (because they use the `Configuration` you changed) and which tests aren't (because they use one of the `Configuration`s that still use data flow rather than taint tracking).

Good point. I started writing an explanation for which configurations affect each test, but it turned out to be much easier to just convert them all to use the new taint tracking configuration.

File: cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_sinks_only/defaulttainttracking.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_sinks_only/defaulttainttracking.cpp:22]: I hope so! I haven't quite understood why we lose this result yet. But since we catch the flow to `buf` a few lines down I'm assuming that there's a simple fix for it (although I spend some time this afternoon looking for that fix, but didn't find it).

File: cpp/ql/test/library-tests/dataflow/taint-tests/stringstream.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/taint-tests/stringstream.cpp]: Fixed in dd6b27df24180f7bcd0e40fc1ed996c301b451b5.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:96]: I expected that all calls to `commonTaintStep` could be removed when you converted the last configurations to use `TaintTracking`. Or at least the body of `commonTaintStep` could cut down to contain just some subset of what's now in `AdjustedConfiguration::isAdditionalTaintStep`.

Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll:96]: Oh, good point. They should all be removed, indeed.

File: cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_sinks_only/defaulttainttracking.cpp
Comment (MathiasVP) [reply cpp/ql/test/library-tests/dataflow/DefaultTaintTracking/annotate_sinks_only/defaulttainttracking.cpp:22]: Oh! Now I know why we lose this result. It's due to the extra sanitizers that were added in an early state of this PR: https://github.com/github/codeql/commit/fbe857d1fa8a3b577e043a74155b804214522373#diff-ab19ecc0bd6242557ac1c5c146893c5295950d847c2d3b216391712bbb5f76eeR224. I think we should accept this lost result for now and revisit it later.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/TaintTrackingUtil.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/TaintTrackingUtil.qll:77]: I think it can happen even without a function call boundary:
```
r0 = ...
r1 = FieldAddress[f1] : r0
r2 = load : r1, ~m1
```
if we have flow from `r0` to `r1` then we'll have flow from `r0` to `r2` due to the `instrTo.(LoadInstruction).getSourceAddressOperand() = opFrom` rule.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll]: Fixed in 3914a93504bb92bb514fe74cb609662f683f290b.

**Comment by MathiasVP** [comment]
I believe I've addressed all comments now. I've started another round of [CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1848/) to check that I didn't mess up anything in 3914a93504bb92bb514fe74cb609662f683f290b.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll]: :heart_eyes:

**Comment by jbj** [comment]
@MathiasVP https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1848/ failed.

LGTM modulo that. I'll leave the merging to @rdmarsh2 (when CPP-Differences looks good).

**Comment by MathiasVP** [comment]
> @MathiasVP https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1848/ failed.

Thanks for the heads up. Retrying: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1849/

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/TaintTrackingUtil.qll
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/TaintTrackingUtil.qll:77]: I think having flow from `r0` to `r2` is OK by itself - If `r0` is a pointer that's computed using tainted data, then `r2` is a load from that tainted pointer and therefore also tainted. It's only when `r0` is tainted because of a pointer/object conflation that having flow from `r0` to `r2` is problematic.

**Comment by MathiasVP** [comment]
[CPP-differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1849/) came back with good results, but terrible performance. This was because `AdjustedConfiguration` was using the same taint configuration as `ToGlobalVarTaintTrackingCfg`.

I've fixed this in 7ec86b5e7f1314f0649888651daff8df75fd6948. Here's the (hopefully) final CPP-difference to confirm that performance is fixed: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1854/.

**Comment by jbj** [comment]
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1854/ timed out after running out of disk space, but that seems to have been a transient error. The last 7 CPP-Differences jobs have succeeded.

**Comment by MathiasVP** [comment]
> https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1854/ timed out after running out of disk space, but that seems to have been a transient error. The last 7 CPP-Differences jobs have succeeded.

Indeed. https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1866/ just came back with a proper result:

- All the results look like good results (they have all been previously checked)
- The performance has been fixed by 7ec86b5e7f1314f0649888651daff8df75fd6948.

I'll merge in main and this should be good to go. :shipit:

**Review by jbj — APPROVED** [review]
The tests are passing on the internal PR.

**Pull request merged**
